File: crates/lure-core/src/provider/registry.rs
//! 最小 provider registry 与选择顺序。
//!
//! 对齐上游 `nanobot/providers/registry.py` 与 `Config._match_provider` 的**选择
//! 顺序本质**：强制 provider 按名查找；`auto` 时先看 `provider/model` 显式前缀，
//! 再按 registry 顺序做关键字匹配。
//!
//! Phase 4 只收录代表性子集，且不含 config 驱动的 api_key/OAuth/local fallback
//! 与 transcription-only 过滤——这些依赖尚未建模的 `ProvidersConfig`，留待后续
//! （见 upstream-test-ledger）。

/// 一个 provider 规格。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ProviderSpec {
    /// registry 名称。
    pub name: &'static str,
    /// 关键字（用于 auto 模型匹配）。
    pub keywords: &'static [&'static str],
    /// 默认 API base URL。
    pub default_api_base: &'static str,
}

/// 代表性 provider 子集，顺序对齐上游 registry 的相对次序。
pub static PROVIDERS: &[ProviderSpec] = &[
    ProviderSpec {
        name: "openrouter",
        keywords: &["openrouter"],
        default_api_base: "https://openrouter.ai/api/v1",
    },
    ProviderSpec {
        name: "anthropic",
        keywords: &["anthropic", "claude"],
        default_api_base: "https://api.anthropic.com/v1",
    },
    ProviderSpec {
        name: "openai",
        keywords: &["openai", "gpt"],
        default_api_base: "https://api.openai.com/v1",
    },
    ProviderSpec {
        name: "deepseek",
        keywords: &["deepseek"],
        default_api_base: "https://api.deepseek.com",
    },
    ProviderSpec {
        name: "gemini",
        keywords: &["gemini", "gemma"],
        default_api_base: "https://generativelanguage.googleapis.com/v1beta/openai/",
    },
    ProviderSpec {
        name: "moonshot",
        keywords: &["moonshot", "kimi"],
        default_api_base: "https://api.moonshot.ai/v1",
    },
    ProviderSpec {
        name: "mistral",
        keywords: &["mistral", "magistral", "ministral", "codestral", "devstral"],
        default_api_base: "https://api.mistral.ai/v1",
    },
    ProviderSpec {
        name: "groq",
        keywords: &["groq"],
        default_api_base: "https://api.groq.com/openai/v1",
    },
    ProviderSpec {
        name: "novita",
        keywords: &["novita"],
        default_api_base: "https://api.novita.ai/openai",
    },
];

/// 按名精确查找 provider（大小写、连字符/下划线归一）。
pub fn find_by_name(name: &str) -> Option<&'static ProviderSpec> {
    let normalized = normalize(name);
    PROVIDERS
        .iter()
        .find(|spec| normalize(spec.name) == normalized)
}

/// 解析模型应使用的 provider。
///
/// - `forced != "auto"`：按名查找。
/// - `auto`：先看 `provider/model` 显式前缀，再按 registry 顺序关键字匹配。
pub fn match_provider(model: &str, forced: &str) -> Option<&'static ProviderSpec> {
    if forced != "auto" {
        return find_by_name(forced);
    }

    if let Some((prefix, _)) = model.split_once('/') {
        if let Some(spec) = find_by_name(prefix) {
            return Some(spec);
        }
    }

    let lower = model.to_lowercase();
    PROVIDERS
        .iter()
        .find(|spec| spec.keywords.iter().any(|kw| lower.contains(kw)))
}

fn normalize(value: &str) -> String {
    value.to_lowercase().replace('-', "_")
}

File: packages/core/src/config/registry.ts
//! 最小 provider registry 与选择顺序（对齐 crates/lure-core/src/provider/registry.rs）。

export interface ProviderSpec {
  name: string;
  keywords: string[];
  defaultApiBase: string;
}

/// 代表性 provider 子集，顺序对齐上游 registry 的相对次序。
export const PROVIDERS: ProviderSpec[] = [
  { name: "openrouter", keywords: ["openrouter"], defaultApiBase: "https://openrouter.ai/api/v1" },
  { name: "anthropic", keywords: ["anthropic", "claude"], defaultApiBase: "https://api.anthropic.com/v1" },
  { name: "openai", keywords: ["openai", "gpt"], defaultApiBase: "https://api.openai.com/v1" },
  { name: "deepseek", keywords: ["deepseek"], defaultApiBase: "https://api.deepseek.com" },
  { name: "gemini", keywords: ["gemini", "gemma"], defaultApiBase: "https://generativelanguage.googleapis.com/v1beta/openai/" },
  { name: "moonshot", keywords: ["moonshot", "kimi"], defaultApiBase: "https://api.moonshot.ai/v1" },
  { name: "mistral", keywords: ["mistral", "magistral", "ministral", "codestral", "devstral"], defaultApiBase: "https://api.mistral.ai/v1" },
  { name: "groq", keywords: ["groq"], defaultApiBase: "https://api.groq.com/openai/v1" },
  { name: "novita", keywords: ["novita"], defaultApiBase: "https://api.novita.ai/openai" },
];

function normalize(value: string): string {
  return value.toLowerCase().replace(/-/g, "_");
}

/// 按名精确查找 provider（大小写、连字符/下划线归一）。
export function findByName(name: string): ProviderSpec | undefined {
  const normalized = normalize(name);
  return PROVIDERS.find((spec) => normalize(spec.name) === normalized);
}

/// 解析模型应使用的 provider（不含 config 驱动的启用/ api_base，见 resolveProvider）。
export function matchProvider(model: string, forced: string): ProviderSpec | undefined {
  if (forced !== "auto") return findByName(forced);

  const slash = model.indexOf("/");
  if (slash >= 0) {
    const prefix = findByName(model.slice(0, slash));
    if (prefix) return prefix;
  }

  const lower = model.toLowerCase();
  return PROVIDERS.find((spec) => spec.keywords.some((kw) => lower.includes(kw)));
}
